### ISSUE TITLE
Add support for address salvage that uses heuristics to salvage a bad address

### ISSUE THREAD

**Issue by mraross** [bcgov/ols-geocoder#14]
Here is a couple of addresses that score poorly in geocoder/addresses:
`2187 SOUTH ALDER ST S, CAMPBELL RIVER, BC`
`MAYFAIR MALL 3147 DOUGLAS ST, VICTORIA, BC`

The first has a redundant streetDirection, the second an irrelevant building name. These examples look salvagable.

Heuristics aren't guaranteed to succeed and they might take much longer than tighter parse rules. This makes heuristics unsuitable for incorporation into the standard parsing process.  We could support a salvage flag in the addresses resource or create a separate resource called addresses/salvage.

The need for salvaging comes from the MOH experience of geocoding all 4.8 million MSP addresses and getting a 6% rejection rate. How long does it take to manually clean up 300,000 addresses? At between one minute and 4 minutes per address,  it could take between 5,000 and 20,000 person hours. Implementing automated address salvage could cut this time in half or more.

**Comment by cmhodgson** [comment]
Note that the c/o case is also handled by #11.

**Comment by cmhodgson** [comment]
A useful option might be to automatically apply the more aggressive and time consuming salvage heuristics on any input address that doesn't have a match within the tolerance. This would allow for a single-pass batch geocoding that takes more time as needed to try to find an acceptable match, rather than requiring users to re-submit failures to a second, more aggressive salvage service. This would be enabled with an additional parameter eg. &tryHard=true

**Comment by mraross** [comment]
Agreed as per my comment on Feb 8 above:

"We could support a salvage flag in the addresses resource or create a separate resource called addresses/salvage"

**Comment by cmhodgson** [comment]
The estimate for this is dependent on how complicated the heuristics we decide to implement are. Some relatively simple examples have been given which may be easily implemented, however the goal will likely be to ensure handling of all "common" failures, as identified by #27, the result of which is not known at this time. 1 week is a reasonable starting point, but we could potentially spend double that to try to catch more cases.

**Comment by mraross** [comment]
I would like to put more time into this and have the work determined by the outcome of the data pattern analysis task so maybe two weeks work?

**Comment by cmhodgson** [comment]
Right, but I am suggesting the salvage operation is conditional on a normal geocode failing; rather than always looking harder for the best result. Depending on the heuristics employed this might be implied (why keep looking if you've already found something "good enough") however the expectation of a "pure" salvage operation is that the input has already been deemed to be hard to match, whereas the 1-pass approach doesn't assume anything about the input.

**Comment by mraross** [comment]
Sorry, you're right. I was assuming salvage=true meant all input addresses were bad.  As a parameter name, neither salvage or tryHard captures the conditionality. Maybe the name doesn't have to as long as we define what it means as in:

salvage - if True, geocoder will attempt a speedy match; if that fails, it will try progressively longer matching heuristics to come up with a plausible match.

**Comment by mraross** [comment]
Here are some unexpected/noise words and phrases found in rejected addresses:

Phone
No Phone
AT - bad abbreviation for apartment
unknown
canada - following st, not province
rtd mail
return mail
mailing

**Comment by mraross** [comment]
Here are some unrecognized building names. They may appear at the beginning or following street. Counts are shown in round brackets:

amica (64) 
aqua (5)
espana + {1,2} (13)
oscar (9) 

name + "apts" (205) - following street or at beginning
apartment building (687) - at beginning or following street
Hotel (9) - as last word in a building name
House (297) - as last word in a building name
Lodge (89) - as last word in a building name
Manor (150) - as last word in a building name
Care Centre or Care Cen (74) - as last words in a building name
sunrise or sunrise of vancouver (62)

**Comment by mraross** [comment]
Noise immunity to unknown site names or other garbage at the front of an address seems to be the most promising thing to implement.  As discussed, you could try the following:

1. In a left-to-right word scan, look for first number token, slap a frontGate just before it, and try to geocode the result; if you get a score of 95 or higher, return the result with an added a SITE_NAME.unknown fault with a five-point penalty. One additional refinement is to check that the word immediately left of the first number token is a unitDescriptor and act appropriately. 

2. If step 1 failed, try a sliding frontGate where place a frontGate after the first word, and try to geocode that; if it returns a perfect match at any given matchPrecision, return the result with an additional five-point penalty. If no perfect match, slide the frontGate right one word and repeat. For example:

   Harry's Harpoonery Atlin, BC

should return the following:

   Harry's Harpoonery -- Atlin, BC

with a score of 63, matchPrecision of locality, and penalty of 5 points for missing frontGate(?).

**Comment by mraross** [comment]
Another possible metric is a strict pattern match with two noise word areas:

<civicNumber> <sw-streetName> <streetType> + <sw-locality> BC
<civicNumber> <sw-streetName> <streetType> <unitDesignator> <unitNumber> + <sw-locality> BC
<unitNumber> <civicNumber> <sw-streetName> <streetType> + <sw-locality> BC

where sw means single-word (e.g. single-word streetName)
           * means one or more noise words

These three patterns should match thousands of addresses with noise.
